### PR TITLE
feat: display address on ledger

### PR DESCRIPTION
**Pull request by obiyankenobi**
Address requested by the wallet is displayed on Ledger and waits for user confirmation.

Wallet sends the address index as a 4-byte unsigned integer (big-endian).

### CODE REVIEW

File: src/getAddress.c [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+// Preprocessor for this screen. Hides left or right arrows depending on the
+// scrolling position.
+static const bagl_element_t* ui_prepro_getAddress_compare(const bagl_element_t *element) {
+    switch (element->component.userid) {
+    case 1:
+        // 0x01 is the left icon so return NULL if we're displaying the beginning of the text.
+        return (ctx->displayIndex == 0) ? NULL : element;
+    case 2:
+        // 0x02 is the right, so return NULL if we're displaying the end of the text.
+        return (ctx->displayIndex == sizeof(ctx->b58_address)-12) ? NULL : element;
```
Comment (msbrogli) [inline src/getAddress.c]: Why do you subtract 12? Is it the screen size? Maybe we should move it to a constant.

File: src/getAddress.c [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+static unsigned int ui_getAddress_compare_button(unsigned int button_mask, unsigned int button_mask_counter) {
+    switch (button_mask) {
+    // scroll left by either clicking or pressing the left button
+    case BUTTON_LEFT:
+    case BUTTON_EVT_FAST | BUTTON_LEFT: // SEEK LEFT
+        // Decrement the displayIndex when the left button is pressed (or held).
+        if (ctx->displayIndex > 0) {
+            ctx->displayIndex--;
+        }
+        os_memmove(ctx->partialAddress, ctx->b58_address+ctx->displayIndex, 12);
```
Comment (msbrogli) [inline src/getAddress.c]: Again the magic number 12. Is it safe to just run `os_memmove` or can it access memory after `ctx->b58_address`? Maybe we should include an overflow protection.

File: src/getAddress.c [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+        UX_REDISPLAY();
+        break;
+
+    // scroll right by either clicking or pressing the right button
+    case BUTTON_RIGHT:
+    case BUTTON_EVT_FAST | BUTTON_RIGHT: // SEEK RIGHT
+        if (ctx->displayIndex < sizeof(ctx->b58_address)-12) {
```
Comment (msbrogli) [inline src/getAddress.c]: 12 magic number again.

File: src/getAddress.c [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+            ctx->displayIndex++;
+        }
+        os_memmove(ctx->partialAddress, ctx->b58_address+ctx->displayIndex, 12);
```
Comment (msbrogli) [inline src/getAddress.c]: 12 magic number again.

File: src/getAddress.c [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+    explicit_bzero(&public_key, sizeof(public_key));
+
+    // convert to base58
+    if (encode_base58(bin_address, sizeof(bin_address), ctx->b58_address, sizeof(ctx->b58_address)) == -1) {
+        // there's been an error
+        THROW(SW_DEVELOPER_ERR);
+    }
+
+    // move the first 12 characters into the partialAddress buffer.
+    os_memmove(ctx->partialAddress, ctx->b58_address, 12);
```
Comment (msbrogli) [inline src/getAddress.c]: 12 magic number again.

File: src/hathor.c [hunk 6]
```diff
@@ -11,33 +11,44 @@
 #include <cx.h>
 #include "hathor.h"
 
+#define B58_MAX_INPUT_SIZE 120
+
+// All keys that we derive start with path 44'/280'/0'
+const uint32_t htr_bip44[] = { 44 | 0x80000000, 280 | 0x80000000, 0 | 0x80000000 };
```
Comment (msbrogli) [inline src/hathor.c]: Is the bit `0x80000000` for hardened keys? Maybe we should leave a comment about it.

File: src/ux.h [hunk 7]
```diff
@@ -5,6 +5,26 @@
...
 
+typedef struct {
+    // public key index for the address
+    uint32_t keyIndex;
+    // base-58 address
+    uint8_t b58_address[34];
+    // the starting index to be shown
+    uint8_t displayIndex;
+    // NULL-terminated string for display
+    uint8_t partialAddress[13];
```
Comment (msbrogli) [inline src/ux.h]: Is it the display size (12) plus one for null termination?

Comment (obiyankenobi) [reply src/ux.h]: Yep.

File: src/getAddress.c (hunk 1 above)
Comment (obiyankenobi) [reply src/getAddress.c]: Yes, it's the maximum string size we show on the screen.

I'll move it to a `#define` on one of the header files.

File: src/getAddress.c (hunk 2 above)
Comment (obiyankenobi) [reply src/getAddress.c]: As in any of the `mem*` functions, an overflow is possible. In this case, we are just copying the 12 bytes and we know that `ctx->partialAddress` can support it.

File: src/hathor.c (hunk 6 above)
Comment (obiyankenobi) [reply src/hathor.c]: Yes, it's specified in BIP32.

Comment (msbrogli) [reply src/hathor.c]: I didn't know that. Thanks!

Even though, maybe we should leave a comment with a link to BIP32 specification. What do you think?

File: src/hathor.c [hunk 8]
```diff
@@ -70,3 +91,65 @@ void hash160(unsigned char *in, unsigned short inlen, unsigned char *out) {
...
 }
+
+int encode_base58(const unsigned char *in, size_t inlen, unsigned char *out, size_t outlen) {
+    unsigned char buffer[B58_MAX_INPUT_SIZE * 138 / 100 + 1] = {0};
+    size_t i = 0, j;
+    size_t startAt, stopAt;
+    size_t zeroCount = 0;
+    size_t outputSize;
+
+    if (inlen > B58_MAX_INPUT_SIZE) {
```
Comment (pedroferreira1) [inline src/hathor.c:103]: Is it maximum size, or we could validate the exact size of the address?

Or is this just a generic method to encode base58 strings without validating addresses?

Comment (obiyankenobi) [reply src/hathor.c:103]: This is a generic method. There's no address validation.

**Review by pedroferreira1 — APPROVED** [review]

File: src/hathor.c (hunk 6 above)
Comment (obiyankenobi) [reply src/hathor.c]: Added the comment.

File: src/getAddress.c (hunk 1 above)
Comment (obiyankenobi) [reply src/getAddress.c]: Defined this on ux.h

File: src/ux.h (hunk 7 above)
Comment (obiyankenobi) [reply src/ux.h]: Used the defined value + 1

**Review by msbrogli — APPROVED** [review]

**Pull request merged**